marimo/_runtime/state.py

```diff
@@ -2,22 +2,108 @@
 from __future__ import annotations
 
 import types
-from typing import Callable, Generic, TypeVar
+import weakref
+from dataclasses import dataclass
+from typing import Any, Callable, Generic, Optional, TypeVar
+from uuid import uuid4
 
 from marimo._output.rich_help import mddoc
 from marimo._runtime.context import ContextNotInitializedError, get_context
 
 T = TypeVar("T")
 
 
+@dataclass
+class StateItem(Generic[T]):
+    id: int
+    ref: weakref.ref[State[T]]
+
+
+class StateRegistry:
+    _states: dict[str, StateItem[Any]] = {}
+    _inv_states: dict[int, set[str]] = {}
+
+    def register(self, state: State[T], name: Optional[str] = None) -> None:
+        if name is None:
+            name = str(uuid4())
+        if id(state) in self._inv_states:
```

Comment:
FYI I have had problems with using Python id as a unique key, because Python reuses ids. Not sure if this would be a problem here.

Reply:
Left a comment, I think there's aggressive enough cleanup